File: go/p40.go
package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	resultList := make([][]int, 0)
	dfs2(candidates, target, 0, make([]int, 0), &resultList)
	return resultList
}

func dfs2(candidates []int, target int, start int, valueList []int, resultList *[][]int) {
	length := len(candidates)
	if target == 0 {
		tmpCurComb := make([]int, len(valueList))
		copy(tmpCurComb, valueList)
		*resultList = append(*resultList, tmpCurComb)
		return
	}
	for i := start; i < length; i++ {
		if target < candidates[i] {
			break
		}
		if i > start && candidates[i] == candidates[i-1] {
			continue
		}
		dfs2(candidates, target-candidates[i], i+1, append(valueList, candidates[i]), resultList)
	}
}

func main() {
	resultList := combinationSum2([]int{10, 1, 2, 7, 6, 1, 5}, 8)
	for _, s := range resultList {
		for _, v := range s {
			print(v)
		}
		println()
	}
}

File: python/p40.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates = sorted(candidates)
        Solution.ret = []
        self.dfs(candidates, target, 0, [])
        return Solution.ret

    def dfs(self, candidates, target, start, value_list):
        length = len(candidates)
        if target == 0:
            Solution.ret.append(value_list)
        for i in range(start, length):
            if target < candidates[i]:
                return
            if i > start and candidates[i] == candidates[i - 1]:
                continue
            self.dfs(candidates, target - candidates[i], i + 1, value_list + [candidates[i]])

if __name__ == '__main__':
    print(Solution().combinationSum2([10, 1, 2, 7, 6, 1, 5], 8))